### PR TITLE
ENH/ BUG: Pydata sparse equality

### LINKED ISSUE #11163: Comparisons between scipy spmatrix and can sparse.SparseArray return scalar bool
This issue is about interop with scipy sparse and pydata sparse. `scipy.sparse` doesn't let `pydata/sparse` overload `__eq__` according to https://github.com/pydata/sparse/issues/305#issuecomment-560317183 (I'm still a little fuzzy on when `__array_priority__` and `__array_ufunc__` get used). This ends up returning a scalar for `!=` or `==`.

#### Reproducing code example:

```python
In [1]: import sparse                                                                               
sra 
In [2]: import scipy                                                                                

In [3]: scipy.__version__                                                                           
Out[3]: '1.5.0.dev0+5958e9f'

In [4]: coo = sparse.random((10, 10))                                                               

In [5]: csr = coo.tocsr()                                                                           

In [6]: csr == coo                                                                                  
Out[6]: False
```

I'd expected this to broadcast and not be False, especially since interop seems to be a goal (#10901).

### PR DESCRIPTION
**Pull request by ivirshup**
#### Reference issue

Fixes #11163

#### What does this implement/fix?

Fix comparison between pydata sparse arrays and sparse matrices.

#### Additional information

I was a little unsure of the best way to do this. I've made a kinda messy function messier, but I don't think this is worth a rewrite of that function. A more elegant solution could be to try and defer to the other object if it satisfies the array interface.

Additionally the test probably don't belong in under `sparse.linalg`, but I wasn't sure where else this should go, especially since I'd be copying code otherwise. I figured I'd leave code organization decisions up to maintainers, so let me know where to move things.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
I think if PyData/Sparse matrices are present, we should just dispatch to that instead, otherwise the well-defined type hierarchy is broken. Adding `if is_pydata_sparse(other): perform_reverse_op(other, self)` should be enough.

**Comment by hameerabbasi** [comment]
Or even `if is_pydata_sparse(other): return NotImplemented` would work.

**Comment by ivirshup** [comment]
Which cases would you expect to work with dispatching to pydata sparse? Right now I think believe testing for all scipy sparse classes.

Also, performance seems ~10x lower with pydata/sparse. Should it still dispatch to that?

<details>
<summary> quick benchmark </summary>

```python
In [10]: coo = sparse.random((1000, 1000))                                                          

In [11]: csr = coo.tocsr()   

In [12]: %timeit coo != csr                                                                         
5.03 ms ± 22.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [13]: %timeit coo != csr                                                                         
5 ms ± 10.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [14]: %timeit csr != coo                                                                         
438 µs ± 4.73 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

</details>

**Comment by hameerabbasi** [comment]
> Which cases would you expect to work with dispatching to pydata sparse? Right now I think believe testing for all scipy sparse classes.

All of them. 😄 

> Also, performance seems ~10x lower with pydata/sparse. Should it still dispatch to that?

Some of my cycles are expected to free up in January for work on PyData/Sparse. Then I'll work on it properly.

**Comment by ivirshup** [comment]
> Some of my cycles are expected to free up in January for work on PyData/Sparse. Then I'll work on it properly.

:+1:

----------

Should it return NotImplemented for pydata sparse exclusively or anything that has `__array_function__` / `__array_ufunc__` and hasn't been special cased already?

**Comment by hameerabbasi** [comment]
> Should it return NotImplemented for pydata sparse exclusively or anything that has __array_function__ / __array_ufunc__ and hasn't been special cased already?

I'd say keeping behavior for other things is probably desirable, otherwise it'd be a large design change. But it should be done for all operators and all formats (and I know that's a big task, so it doesn't have to be you or this PR).

**Comment by ivirshup** [comment]
Yeah, I might leave that to someone else/ at least until the morning. `__eq__` and `__ne__` are deferring to pydata/sparse now.

**Review by hameerabbasi — APPROVED** [review]

File: scipy/sparse/linalg/tests/test_pydata_sparse.py [hunk 1]
```diff
@@ -202,3 +223,13 @@ def test_expm_multiply(matrices):
     x0 = splin.expm_multiply(A_dense, b)
     x = splin.expm_multiply(A_sparse, b)
     assert_allclose(x, x0)
+
+
+def test_eq(same_matrix):
+    sp_sparse, pd_sparse = same_matrix
+    assert np.all(sp_sparse == pd_sparse)
```
Comment (hameerabbasi) [inline scipy/sparse/linalg/tests/test_pydata_sparse.py]:
```suggestion
    assert (sp_sparse == pd_sparse).all()
```

File: scipy/sparse/linalg/tests/test_pydata_sparse.py [hunk 2]
```diff
@@ -202,3 +223,13 @@ def test_expm_multiply(matrices):
...
+
+
+def test_ne(same_matrix):
+    sp_sparse, pd_sparse = same_matrix
+    assert not np.any(sp_sparse != pd_sparse)
```
Comment (hameerabbasi) [inline scipy/sparse/linalg/tests/test_pydata_sparse.py]:
```suggestion
    assert not (sp_sparse != pd_sparse).any()
```

**Review by hameerabbasi — COMMENTED** [review]
`np.any` and `np.all` require `__array_function__`, IIUC.

**Comment by perimosocordiae** [comment]
I'm traveling today, but I'm marking myself on this PR so I won't forget to come back to it.

**Comment by ivirshup** [comment]
> np.any and np.all require `__array_function__`, IIUC.

@hameerabbasi, this would fit my understanding too, but what I've put still seems to work with `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=0`:

```python
import numpy as np
from scipy import sparse as ss
import sparse
import os

os.environ["NUMPY_EXPERIMENTAL_ARRAY_FUNCTION"]
# '0'

coo = sparse.random((100, 100))
csr = coo.tocsr()

csr == coo
# <COO: shape=(100, 100), dtype=bool, nnz=0, fill_value=True>

not np.any(csr != coo)
# True

not np.all(csr == coo)
# False

np.all(csr == coo)
# True
```

**Comment by hameerabbasi** [comment]
> @hameerabbasi, this would fit my understanding too, but what I've put still seems to work with `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=0`:

Ahh, yes. It uses the `.all()` and `.any()` methods.

**Comment by pv** [comment]
Thanks, seems ok to merge.

There's indeed bigger overhaul required vs `array_ufunc` / `array_priority`, but the backward compat issues are annoying there, which is why it's not been done

**Comment by mdhaber** [comment]
Thanks @ivirshup ! If you'd like your name added to [THANKS.txt](https://github.com/scipy/scipy/blob/master/THANKS.txt), please make an entry [here](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*).

**Pull request merged**